src/ensembl/src/store.ts

```diff
@@ -4,6 +4,41 @@ import { composeWithDevTools } from 'redux-devtools-extension';
 
 import reducers from './rootReducer';
 
+import ReactGA from 'react-ga';
+
+ReactGA.initialize('UA-58710484-17');
+
 const composeEnhancers = composeWithDevTools({});
 
-export default createStore(reducers, composeEnhancers(applyMiddleware(thunk)));
+const googleAnalyticsMiddleWare = (store: any) => (next: any) => (
+  action: any
+) => {
+  try {
+    // If we have the google anlytics meta data passed in, and
+    // if disable_ga_tracking flag is not set to true,
+    // We need to track this event
+    if (!action.meta.ga.disable_tracking && action.meta.ga.category.length) {
```

Comment:
Will action always have the `meta` field? Will there always be the `ga` object on the `meta` field?

Reply:
This is the way i'm trying to enforce the developer to include the meta information every time they implement a new action. If the meta or meta.ga is missing, a warning message will be displayed on the console.

Reply:
How will it work with e.g. thunks? Wouldn't all of them cause errors?

Reply:
It will cause error as long as the meta is not defined. I couldn't think of any workarounds other than passing the 'disable_tracking' flag.

Also, we will be able to track how long an AJAX request took to get the response, for which we will need to pass some meta along with the event dispatched from the thunk. This will be helpful to filter the requests that takes long time.

Reply:
Could we perhaps disable this check then and rely on each other during code review to make sure we are adding analytics metainfo when required?

Reply:
Could we perhaps remove the automatic check then and rely on each other during the code review process to point out that an analytics event is required (when it is)?

Reply:
Why do you suggest removing the automatic check ?That's more robust than code review...

Reply:
> Could we perhaps remove the automatic check then and rely on each other during the code review process to point out that an analytics event is required (when it is)?

I feel it will be a good way to remind the developers to include the meta info as we still in the initial stages of development. It will pretty much reduce the number of checks we need to during the code review process.

Reply:
> Why do you suggest removing the automatic check ?

The check assumes that all our actions will be objects with a `meta` field on them. Thunks are actions that are not objects; they are functions; so I would expect the check to break. Moreover, if we are going to use a more sophisticated redux middleware for side effects (such as `redux-saga` or `redux-observable`), that will result in more actions that will not necessarily have the `meta` field on them, nor should they.

Showing console errors in all these cases will be distracting and once false-positive errors appear, the developer will never pay attention to them.

Reply:
Redux actions is an internal messaging mechanism within the app. They do not correspond 1:1 to user's interactions with the UI.

Reply:
>The check assumes that all our actions will be objects with a meta field on them. Thunks, on the other hand, are actions that are not objects; they are functions; so I would expect the check to break.

As far as I know, all the actions must be of type object which is explained in the links:
- https://redux.js.org/basics/actions#actions
- https://github.com/redux-utilities/flux-standard-action#actions

If am not wrong, the way thunk works is that it will initially return a function that eventually dispatches an action, which is again of type object. The function returned by thunk will be handled by thunk itself and not the middlewares. 

If the above statement is right, it should all be the same for thunk, saga and observable where we will only be handling the final action, which will be an object.

In the second link I have included above, it states that all the actions MUST include the 'type' property and MAY have the meta property, which is what I am trying to make use of to implement this tracking.

> Showing console errors in all these cases will be distracting, and once false-positive errors appear, the developer will no longer pay attention to them.

I agree that false-positive errors have other side effects as you have mentioned, but I've used console.warn which instead gets displayed as a warning message in amber. 

> Redux actions are an internal messaging mechanism within the app. They do not correspond 1:1 to user's interactions with the UI.

This is true, but again, this mechanism already has pretty much everything we needed for google analytics tracking, using which we can avoid building and running another similar mechanism in parallel.

Reply:
> As far as I know, all the actions must be of type object which is explained in the links:

That’s the general rule, yes, but thunks break this contract — see their implementation here:
https://github.com/reduxjs/redux-thunk/blob/master/src/index.js#L1-L9

Notice also that (if I am reading the source code correctly) these functions will get propagated through the chain of middlewares (thunk middleware calls `next(action)` and passes this action function into `next`).

Thunks do not observe the shape of a Flux standard action. And I am sure we will be using them. It is also not a far stretch to imagine that other redux middleware we might add will not need the `meta` field on the action object.

Speaking of Flux standard actions, raising an error (even though its manifestation is a warning text in the console) when the `meta` field is absent is essentially breaking with the FSA standard, because, as you rightly quoted, the `meta` field is optional.

> this mechanism already has pretty much everything we needed for google analytics tracking

The mechanism is fine; let’s just not produce errors when the action lacks the `meta` field. This puts us in the position when we either have to ensure that all actions, even for internal workings of the app, have an analytics object attached to them (which may not even be desirable), or have to accept false-positives (and thus lose the usefulness of the notification). Let’s have the freedom of adding analytics objects only to the actions that are relevant for analytics.

Reply:
> (thunk middleware calls next(action) and passes this action function into next).

I think the next(action) will be called only when the action is not of type function. If it is of type function, it dispatches another action which again gets handled by Thunk. When this Thunk gets resolved, the final dispatch of action will be an object, which gets passed on to our custom analytics middleware. This is where we will actually be looking for the `meta` and all the previous actions does not require it.

> Speaking of Flux standard actions, raising an error (even though its manifestation is a warning text in the console) when the meta field is absent is essentially breaking with the FSA standard, because, as you rightly quoted, the meta field is optional, and the error implies that it's obligatory.

It does break the FSA standard, but I was thinking about having a standard of our own in order to solve the requirement of tracking almost all the events. 


> The mechanism is fine; let’s just not produce errors when the action lacks the meta field. This puts us in the position when we either have to ensure that all actions, even for internal workings of the app, have an analytics object attached to them (which may not even be desirable), or have to accept false-positives (and thus lose the usefulness of the notification).

At the moment, this automatic enforcement is only applied to all the actions that affects the root state, which I think it would all be worthy enough to be tracked. The idea behind displaying the warning on the console is just to make sure that its addressed during the development phase.
I believe that the actions involved in internal workings would mostly not affect the global state like the accordion toggle for example, which we don't have to worry about adding the `meta` info.


> Let’s have the freedom of adding analytics objects only to the actions that are relevant for analytics.
 
If we are okay to make a list of all the actions that needs to be tracked and also rely on manually ensuring that the tracking meta data is correctly added to the new actions during the code review stage, I am happy to remove the warnings. I would suggest to keep it for a while and then we can always scrap it when we get to a point where it's gets very less helpful.

Reply:
> I think the next(action) will be called only when the action is not of type function.

Oh, right :smile:

Reply:
> If we are okay to make a list of all the actions that needs to be tracked and also rely on manually ensuring that the tracking meta data is correctly added to the new actions during the code review stage, I am happy to remove the warnings.

How about using a prefix for all action types that need to be tracked? We can then do this check for those specific actions.

Reply:
We can do that but it wouldn't give us a way enforce the tracking data when it's necessary. Let's say after a year, we will not be very much aware that certain actions needs to be tracked while developing and we could end up adding an entire feature without adding any tracking data to it.